strawberryfields/api/devicespec.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a class that represents the specifications of
+a device available via the API.
+"""
+from collections.abc import Sequence
+
+import strawberryfields as sf
+import blackbird
+
+from strawberryfields.circuitspecs import Ranges
+
+
+class DeviceSpec:
+    """The specifications for a specific hardware device.
+
+    Args:
+        target (str): name of the target hardware device
+        layout (str): string containing the Blackbird circuit layout
+        modes (int): number of modes supported by the target
+        compiler (list): list of supported compilers
+        gate_parameters (dict): parameters for the circuit gates
+        connection (strawberryfields.api.Connection): connection over which the
+            job is managed
+    """
+
+    def __init__(self, target, layout, modes, compiler, gate_parameters, connection):
+        self._target = target
+        self._layout = layout
+        self._modes = modes
+        self._compiler = compiler
+        self._gate_parameters = gate_parameters
+
+        self._connection = connection
+
+    @property
+    def target(self):
+        """str: The name of the target hardware device."""
+        return self._target
+
+    @property
+    def layout(self):
+        """str: Returns a string containing the Blackbird circuit layout."""
+        return self._layout
+
+    @property
+    def modes(self):
+        """int: Number of modes supported by the device."""
+        return self._modes
+
+    @property
+    def compiler(self):
+        """list[str]: A list of strings corresponding to Strawberry Fields compilers supported
+        by the hardware device."""
+        return self._compiler
+
+    @property
+    def gate_parameters(self):
+        """dict[str, strawberryfields.circuitspecs.Ranges]: A dictionary of gate parameters
+        and allowed ranges.
+
+        The parameter names correspond to those present in the Blackbird circuit layout.
+
+        **Example**
+
+        >>> spec.gate_parameters
+        {'squeezing_amplitude_0': x=0, x=1, 'phase_0': x=0, 0≤x≤6.283185307179586}
```

Comment:
(Out of scope): find `'squeezing_amplitude_0': x=0, x=1` slightly confusing, would perhaps prefer `'squeezing_amplitude_0': x ∈ {0, 1}`. For intervals this would get slightly more complicated, having to have unions of sets with `U`. Could just be me.

Reply:
Agreed! We're thinking about updating the `Ranges` class to produce nicer/clearer outputs (it's quite messy right now since the actual output would be `{'squeezing_amplitude_0': squeezing_amplitude_0=0, squeezing_amplitude_0=1, 'phase_0': phase_0=0, 0≤phase_0≤6.283185307179586}` in this case. :confused: 

This would be in a separate PR, though, and it could also solve the issue you mentioned above with the ordering.

Reply:
Yep, let's save this for a subsequent PR